src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java

```diff
@@ -0,0 +1,203 @@
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.security.*;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.Map;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class AttestationTokenRetriever {
+    private static final Logger LOGGER = LoggerFactory.getLogger(UidCoreClient.class);
+    private final IAttestationProvider attestationProvider;
+    private final ApplicationVersion appVersion;
+    private AtomicReference<String> attestationToken;
+    private Handler<Integer> responseWatcher;
+    private final String attestationEndpoint;
+    private final HttpClient httpClient;
+    private final IClock clock;
+    private Vertx vertx;
+    private long timerId;
+    // Set this to be Instant.MAX so that if it's not set it won't trigger the re-attest
+    private Instant attestationTokenExpiresAt = Instant.MAX;
+    private final Lock lock;
+    private final AttestationTokenDecryptor attestationTokenDecryptor;
+
+    private static boolean isGeneratorInitialised = false;
+
+    public AttestationTokenRetriever(String attestationEndpoint, ApplicationVersion appVersion, IAttestationProvider attestationProvider,
+                                     Handler<Integer> responseWatcher, IClock clock, HttpClient httpClient,
+                                     AttestationTokenDecryptor attestationTokenDecryptor) throws IOException {
+        this.attestationEndpoint = attestationEndpoint;
+        this.appVersion = appVersion;
+        this.attestationProvider = attestationProvider;
+        this.attestationToken = new AtomicReference<>(null);
+        this.responseWatcher = responseWatcher;
+        this.clock = clock;
+        this.lock = new ReentrantLock();
+        this.timerId = -1;
+        if (httpClient == null) {
+            this.httpClient = HttpClient.newHttpClient();
+        } else {
+            this.httpClient = httpClient;
+        }
+        if (attestationTokenDecryptor == null) {
+            this.attestationTokenDecryptor = new AttestationTokenDecryptor();
+        } else {
+            this.attestationTokenDecryptor = attestationTokenDecryptor;
+        }
+    }
+
+    private void attestationExpirationCheck(long timerId) {
+        Instant currentTime = clock.now();
+        Instant tenMinutesBeforeExpire = attestationTokenExpiresAt.minusSeconds(600);
+
+        if (currentTime.isAfter(tenMinutesBeforeExpire)) {
+            LOGGER.info("Attestation token is 10 mins from the expiry timestamp %s. Re-attest...", attestationTokenExpiresAt);
+            long oldTimerId = this.timerId;
+            try {
+                attest();
+                this.vertx.cancelTimer(oldTimerId);
```

Comment:
What restarts the timer? The first time we attest, timerId == -1, so we set the timer. On the next attest, the timerId is not -1, so we don't restart it, and we never reset the timerId to -1, so I don't see how it gets scheduled again

Reply:
You are right!!! Fixed